cluster/images/etcd/migrate/data_dir.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+// DataDirectory provides utilities for initializing and backing up an
+// etcd "data-dir" as well as managing a version.txt file to track the
+// etcd server version and storage verion of the etcd data in the
+// directory.
+type DataDirectory struct {
+	path        string
+	dir         *os.File
+	versionFile *VersionFile
+}
+
+// OpenOrCreateDataDirectory opens a data directory, creating the directory
+// if it doesn't not already exist.
+func OpenOrCreateDataDirectory(path string) (*DataDirectory, error) {
+	exists, err := exists(path)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		glog.Infof("data directory '%s' does not exist, creating it", path)
+		err := os.MkdirAll(path, 0755)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create data directory %s: %s", path, err)
+		}
+	}
+	dir, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open data directory %s: %s", path, err)
+	}
+
+	versionFile := &VersionFile{
+		path: filepath.Join(path, versionFilename),
+	}
+	return &DataDirectory{path, dir, versionFile}, nil
+}
+
+// Close releases the underlying File used by the DataDirectory.
+func (d *DataDirectory) Close() error {
+	return d.dir.Close()
+}
+
+// Initialize set the version.txt to the target version if the data
```

Comment:
I think this part correspond to the logic in # If there is no data in DATA_DIRECTORY, this means that we are, I like the smart new way, I am just not sure how obvious the new comment is for users, would everyone know that 1. if the data directory is empty then we don't need to do migration? 2. if version.txt in etcd data folder is set to target version then there is no migration? Maybe it's actually obvious to other people, ignore me then.